File: csharp/873.length-of-longest-fibonacci-subsequence.cs
/*
 * @lc app=leetcode id=873 lang=csharp
 *
 * [873] Length of Longest Fibonacci Subsequence
 */

using System;
using System.Collections.Generic;

// @lc code=start
public class Solution {
    public int LenLongestFibSubseq(int[] A) {
        int n = A.Length, ret = 0;
        int[,] dp = new int[n, n];
        var map = new Dictionary<int, int>();
        for (int i = 0; i < n; ++i) {
            int val = A[i];
            map[val] = i;
            for (int j = i - 1; j >= 0; --j) {
                int diff = val - A[j];
                if (2 * diff < val && map.TryGetValue(diff, out int k)) {
                    dp[i, j] = 1 + dp[j, k];
                    ret = Math.Max(ret, dp[i, j]);
                } else {
                    dp[i, j] = 2;
                }
            }
        }
        return ret > 2 ? ret : 0;
    }
}
// @lc code=end

File: java/873.length-of-longest-fibonacci-subsequence.java
/*
 * @lc app=leetcode id=873 lang=java
 *
 * [873] Length of Longest Fibonacci Subsequence
 */

// @lc code=start
class Solution {
    public int lenLongestFibSubseq(int[] arr) {
        int n = arr.length, ret = 0;
        int[][] dp = new int[n][n];
        Map<Integer, Integer> map = new HashMap<Integer, Integer>();
        for (int j = 0; j < n; ++j) {
            map.put(arr[j], j);
            for (int i = j - 1; i >= 0; --i) {
                int d = arr[j] - arr[i];
                if (d >= arr[i]) {
                    dp[i][j] = 2;
                    continue;
                }
                int p = map.getOrDefault(d, -1);
                if (p < 0) {
                    dp[i][j] = 2;
                } else {
                    dp[i][j] = 1 + dp[p][i];
                    ret = Math.max(ret, dp[i][j]);
                }
            }
        }
        return ret;
    }
}
// @lc code=end

